AK/Optional.h

```diff
@@ -185,6 +185,119 @@ class [[nodiscard]] Optional {
     alignas(T) u8 m_storage[sizeof(T)];
     bool m_has_value { false };
 };
+
+template<typename T>
+class [[nodiscard]] Optional<T&> {
+public:
+    using ValueType = T&;
+
+    ALWAYS_INLINE Optional() = default;
+
+    ALWAYS_INLINE Optional(Optional const& other) = delete;
+    ALWAYS_INLINE Optional(Optional&& other)
+        : m_ptr(other.m_ptr)
+    {
+        other.m_ptr = nullptr;
+    }
+
+    ALWAYS_INLINE Optional(T& value)
+        : m_ptr(&value)
+    {
+    }
+
+    ALWAYS_INLINE ~Optional() = default;
+
+    // Avoid the question of deep-copy vs. shallow-copy by simply deleting copy-assign.
+    // Note that the copy-constructor does a shallow copy, just like T&.
+    ALWAYS_INLINE Optional& operator=(Optional const&) = delete;
+    ALWAYS_INLINE Optional& operator=(Optional&& other)
+    {
+        m_ptr = other.m_ptr;
+        other.m_ptr = nullptr;
+        return *this;
+    }
+
+    ALWAYS_INLINE Optional& operator=(T const&) = delete;
+    ALWAYS_INLINE Optional& operator=(T&&) = delete;
+
+    template<typename O>
+    ALWAYS_INLINE bool operator==(Optional<O> const& other) const
+    {
+        return has_value() == other.has_value() && (!has_value() || value() == other.value());
+    }
+
+    template<typename O>
+    ALWAYS_INLINE bool operator==(O const& other) const
+    {
+        return has_value() && value() == other;
+    }
+
+    ALWAYS_INLINE void clear()
+    {
+        m_ptr = nullptr;
+    }
+
+    ALWAYS_INLINE void emplace(T& parameter)
+    {
+        m_ptr = parameter;
+    }
```

Comment:
I can see that this compiles, but why? How can we assign a reference to a pointer?